[Feature] Allow to handle parameters which have declared as ref

s0lus opened the pull request:
Hey, @ovr!

I want to extend functional for this tool. And i want to add handling of parameters, which declared as`ref`. I have a example to simplify understand for what this feature do.

**TL;DR** Properly handling swagger definitions, which have a following view: ` @SWG\Parameter(ref="#/parameters/fields")`

For example i have definitions for fields `fields`, which i will reuse in other definitions (i.e. in many other places)

```php
/**
 *  @SWG\Parameter(
 *    name="fields",
 *    type="string",
 *    in="query",
 *    description="description"
 *  )
 */
```

Usage:

```php
/**
 * @SWG\Get(
 *  tags={"Tags"},
 *  path="path",
 *  operationId="operationId",
 *  @SWG\Parameter(ref="#/parameters/fields"),
 * )
 */
```

**Expected output will be:**
```js
fields?: string
```

**Actual output:** (because it's not handling properly)
```js
?:
```

Thanks, and have a nice day!

s0lus committed:
[Feature] Allow to handle parameters, which have declared as ref

ovr commented on src/Command/Generate.php:
```diff
@@ -98,11 +106,29 @@ function ($value) {
+                            $stripped = stripParameters($parameter->ref);
+
+                            if ($stripped === $param->name) {
+                                if ($param->type === 'integer') {
```
Тут стоять - смотри в twig что у нас юзается 
```
{{ flowParameterType(parameter)|raw }}
```

flowParameterType - это фильтр - как раз на это дело ;)

https://github.com/SocialConnect/swagger-codegen/blob/45fe1a40a390914ac9b6185be88350eb242fbd73/src/Command/Generate.php#L135

и тут есть case на `integer`

удали пожалуйста if и проверь еще раз

Спасибо!

ovr replied:
Аааа, все понял))) 2дня конференций - сломали меня)

![image](https://user-images.githubusercontent.com/572096/40292134-18765ad0-5cd2-11e8-890c-256f507b169e.png)

А нам стоит использовать - одно ветвление для обоих случаев!

s0lus replied:
@ovr Обновил. https://github.com/SocialConnect/swagger-codegen/pull/2/commits/160a0714343061f878072062fd0c7225bbcb9afa

s0lus committed:
[Feature] Extracts handling of the parameter type in a function (DRY)

ovr commented on src/resource/js/params.twig:
```diff
@@ -2,7 +2,11 @@
 export type {{ operationId }}Params = {
 {% for parameter in parameters %}
 {% if parameter.in != "path" and parameter.in != "body" %}
+{% if parameter.ref is not empty %}
```
@s0lus зачем ты это добавил? У тебя итак flowParameterType - это должен делать!

s0lus replied:
@ovr `flowParameterType`, как я понял, и судя по названию, нужен для обработки типа, а мне нужно имя. В любом случае, там баг был с определением свойства `required` для параметров определенных через `ref`. Поправил в https://github.com/SocialConnect/swagger-codegen/pull/2/commits/8270663a29c5d5b4efe85270faed456ed22f5a48

ovr commented on src/Command/Generate.php:
```diff
@@ -98,25 +127,22 @@ function ($value) {
-                            }
+                    if ($parameter->ref) {
+                        foreach ($swagger->parameters as $param) {
+                            $stripped = stripParameters($parameter->ref);
```
```$stripped = stripParameters($parameter->ref);```
вынеси из цикла - зачем тебе это N раз делать то!

s0lus replied:
@ovr Обновил!

s0lus committed:
[Feature] Introduce flowParameterWithRef

	- Allow to proper handling of the name/required properties for parameters declared with ref keyword

ovr commented:
Thank you 🏅
